provider/hana/hana.go

```diff
@@ -0,0 +1,820 @@
+	ConfigKeySQL             = "sql"
+	ConfigKeyFields          = "fields"
+	ConfigKeyGeomField       = "geometry_fieldname"
+	ConfigKeyFeatureIDField  = "id_fieldname"
+	ConfigKeyGeomType        = "geometry_type"
+	ConfigKeyBuffer          = "buffer"
+	ConfigKeyClipGeometry    = "clip_geometry"
+)
+
+type DataType byte
+
+const (
+	DtTinyint DataType = iota
+	DtSmallint
+	DtInteger
+	DtBigint
+	DtDecimal
+	DtSmalldecimal
+	DtReal
+	DtDouble
+	DtChar
+	DtVarchar
+	DtNChar
+	DtNVarchar
+	DtShorttext
+	DtAlphanum
+	DtBinary
+	DtVarbinary
+	DtDate
+	DtTime
+	DtTimestamp
+	DtSeconddate
+	DtClob
+	DtNClob
+	DtBlob
+	DtText
+	DtBoolean
+	DtSTPoint
+	DtSTGeometry
+	DtUnknown
+)
+
+type FieldDescription struct {
+	dataType    DataType
+	name        string
+	isGeometry  bool
+	isFeatureId bool
+}
+
+func OpenDB(uri string) (*sql.DB, error) {
+	u, err := url.Parse(uri)
+	if err != nil {
+		return nil, err
+	}
+
+	if u.Scheme != "hdb" {
+		return nil, ErrInvalidURI{Msg: fmt.Sprintf("invalid scheme '%v'", u.Scheme)}
+	}
+	params := u.Query()
+
+	supportedParams := []string{ConfigKeyMaxConn, ConfigKeyMaxConnIdleTime, ConfigKeyMaxConnLifetime, driver.DSNTimeout, driver.DSNTLSInsecureSkipVerify, driver.DSNTLSRootCAFile, driver.DSNTLSServerName}
+	for k := range params {
+		found := false
+		for i := 0; i < len(supportedParams); i++ {
+			pname := supportedParams[i]
+			if k == pname {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			return nil, ErrInvalidURI{Msg: fmt.Sprintf("parameter '%v' is unknown", k)}
+		}
+	}
+
+	max_conn := DefaultMaxConn
+	if params.Has(ConfigKeyMaxConn) {
+		max_conn, err = strconv.Atoi(params.Get(ConfigKeyMaxConn))
+		if err != nil {
+			return nil, ErrInvalidURI{Msg: "max_connections value is incorrect"}
+		}
+	}
+
+	max_conn_idle_time, _ := time.ParseDuration(DefaultMaxConnIdleTime)
+	if params.Has(ConfigKeyMaxConnIdleTime) {
+		value := params.Get(ConfigKeyMaxConnIdleTime)
+		max_conn_idle_time, err = time.ParseDuration(value)
+		if err != nil {
+			return nil, ErrInvalidURI{Msg: "max_connection_idle_time value is incorrect"}
+		}
+	}
+
+	max_conn_life_time, _ := time.ParseDuration(DefaultMaxConnLifetime)
+	if params.Has(ConfigKeyMaxConnLifetime) {
+		value := params.Get(ConfigKeyMaxConnLifetime)
+		max_conn_life_time, err = time.ParseDuration(value)
+		if err != nil {
+			return nil, ErrInvalidURI{Msg: "max_connection_life_time value is incorrect"}
+		}
+	}
+
+	// We construct a new uri that only contains parameters supported by the HANA driver.
+	// Otherwise, driver.NewDSNConnector returns an error.
+	newParams := url.Values{}
+	for i := 3; i < len(supportedParams); i++ {
+		pname := supportedParams[i]
+		if params.Has(pname) {
+			value := params.Get(pname)
+			if pname == driver.DSNTLSServerName && value == "host" {
+				value = strings.Split(u.Host, ":")[0]
+			}
+			newParams.Add(pname, value)
+		}
+	}
+
+	newUri := fmt.Sprintf("%v://%v@%v?", u.Scheme, u.User.String(), u.Host) + newParams.Encode()
```

Comment:
consider using `url.URL` to form up the URI. https://pkg.go.dev/net/url#URL

Reply:
I will do so.

Reply:
Thanks!